zipline/algorithm.py

```diff
@@ -632,42 +632,96 @@ def run(self, data=None, overwrite_sim_params=True):
     def _write_and_map_id_index_to_sids(self, identifiers, as_of_date):
         # Build new Assets for identifiers that can't be resolved as
         # sids/Assets
-        identifiers_to_build = set()
-        next_sid = max(self.asset_finder.sids or (0,)) + 1
+        def is_unknown(asset_or_sid):
+            sid = op.index(asset_or_sid)
+            return self.asset_finder.retrieve_asset(
```

Comment:
I don't think it matters enough to be worth the extra complexity. This branch is only hit when you pass in a dataframe or panel as your dataset, which generally means you're running one of our examples, or you've loaded up your data in memory.  In either case, you probably don't have a ton of entries, and if you do have a ton of entries, you'd be much better off writing your data statically to disk.

This code is also all obsoleted by @llllllllll's data bundle work.